central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,623 @@
+	cronJobsLock sync.Mutex
+	// Use to synchronize access to reportsQueue and runningReportConfigs
+	schedulerLock sync.Mutex
+	// Use to lock any database tables if needed to prevent race conditions
+	dbLock sync.Mutex
+	// NOTE: Lock only one mutex at a time. Do not lock another mutex when one is already held.
+	//      If you need to lock another mutex, you must free the one already locked first.
+
+	cron            *cron.Cron
+	concurrencySema *semaphore.Weighted
+}
+
+// ReportRequest is a request to the scheduler to run a scheduled or on demand report
+type ReportRequest struct {
+	ReportConfig   *storage.ReportConfiguration
+	ReportMetadata *storage.ReportMetadata
+	collection     *storage.ResourceCollection
+	dataStartTime  *types.Timestamp
+	Ctx            context.Context
+}
+
+type reportEmailFormat struct {
+	BrandedProductName string
+	WhichVulns         string
+	DateStr            string
+}
+
+// New instantiates a new cron scheduler and supports adding and removing report requests
+func New(reportConfigDatastore reportConfigDS.DataStore, reportMetadataStore reportMetadataDS.DataStore,
+	reportSnapshotStore reportSnapshotDS.DataStore, notifierDatastore notifierDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor) Scheduler {
+	cronScheduler := cron.New()
+	cronScheduler.Start()
+	ourSchema, err := graphql.ParseSchema(resolvers.Schema(), resolvers.New())
+	if err != nil {
+		panic(err)
+	}
+	return newSchedulerImpl(reportConfigDatastore, reportMetadataStore, reportSnapshotStore, notifierDatastore,
+		deploymentDatastore, watchedImageDatastore, collectionDatastore, collectionQueryRes, notificationProcessor, cronScheduler, ourSchema)
+}
+
+func newSchedulerImpl(reportConfigDatastore reportConfigDS.DataStore, reportMetadataStore reportMetadataDS.DataStore,
+	reportSnapshotStore reportSnapshotDS.DataStore, notifierDatastore notifierDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor, cronScheduler *cron.Cron, schema *graphql.Schema) *scheduler {
+	s := &scheduler{
+		reportConfigToEntryIDs:  make(map[string]cron.EntryID),
+		reportConfigDatastore:   reportConfigDatastore,
+		reportMetadataStore:     reportMetadataStore,
+		reportSnapshotStore:     reportSnapshotStore,
+		notifierDatastore:       notifierDatastore,
+		deploymentDatastore:     deploymentDatastore,
+		watchedImageDatastore:   watchedImageDatastore,
+		collectionDatastore:     collectionDatastore,
+		collectionQueryResolver: collectionQueryRes,
+		notificationProcessor:   notificationProcessor,
+		reportsQueue:            make([]*ReportRequest, 0),
+		sendNewReports:          make(chan bool, 200),
```

Comment:
Just a note that if there are pushes more than the buffer size, this will result in fatal error. Likely to happen as reporting framework is integrated into more product workflows.

Reply:
I tested this, if there are more pushes than buffer size, the channel will block the caller until values are read from the channel by another routine. Nevertheless, this was risky so I updated to using signal instead.